python/lsst/afw/table/base.py

```diff
@@ -0,0 +1,219 @@
+from __future__ import absolute_import, division, print_function
+
+from past.builtins import basestring
+
+from .record import addRecordMethods
+from .table import addTableMethods
+from .catalog import addCatalogMethods
+from ._base import BaseRecord, BaseTable, BaseCatalog
+
+import numpy as np
+
+__all__ = []  # import this module only for its side effects
+
+
+def addBaseRecordMethods(cls):
+    """Add pure Python methods to BaseRecord
+    """
+    addRecordMethods(cls)
+
+    def _set(self, key, value):
+        """
+        Given a string or `lsst.afw.table.Key`, find the `BaseRecord` for the given key and set its value.
+
+        The leading underscore avoids shadowing the standard Python `set` class.
+        """
+        if isinstance(key, basestring):
+            try:
+                funcKey = self.getSchema().find(key).key
+            except:
+                raise KeyError("Field '%s' not found in Schema." % key)
+        else:
+            funcKey = key
+
+        try:
+            prefix, suffix = type(funcKey).__name__.split("_")
+        except ValueError:
+            # Try using a functor
+            return key.set(self, value)
+        except Exception:
+            raise TypeError("Argument to BaseRecord.set must be a string or Key.")
+
+        if prefix != "Key":
+            raise TypeError("Argument to BaseRecord.set must be a string or Key.")
+        method = getattr(self, 'set'+suffix)
+        return method(funcKey, value)
+    cls.set = _set
+
+    def __setitem__(self, key, value):
+        """
+        Given a string or `lsst.afw.table.Key`, find the `BaseRecord` for the given key and set its value.
+        """
+        self.set(key, value)
+    cls.__setitem__ = __setitem__
+
+    def get(self, key):
+        """
+        Given a string or `lsst.afw.table.Key`, get the `BaseRecord` for the given key and return it.
+        """
+        if isinstance(key, basestring):
+            try:
+                funcKey = self.getSchema().find(key).key
+            except:
+                raise KeyError("Field '%s' not found in Schema." % key)
+        else:
+            funcKey = key
+
+        try:
+            prefix, suffix = type(funcKey).__name__.split("_")
+        except ValueError:
+            # Try using a functor
+            return key.get(self)
+
+        except Exception:
+            raise TypeError("Argument to BaseRecord.set must be a string or Key.")
+        if prefix != "Key":
+            raise TypeError("Argument to BaseRecord.set must be a string or Key.")
+        method = getattr(self, '_get_'+suffix)
+        return method(funcKey)
+    cls.get = get
```

Comment:
I filed a ticket https://jira.lsstcorp.org/browse/DM-8716 and originally thought @fred3m was going to take it, but he's too busy so it is unassigned.